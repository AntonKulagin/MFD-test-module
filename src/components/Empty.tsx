export const Empty = () => {
  return (
    <div className="empty-wrapper">
      Вы пока не создали ни одной задачи
      <br />
      или все задачи выполнены
      <br />
      Давайте что-нибудь запланируем?
    </div>
  )
}
