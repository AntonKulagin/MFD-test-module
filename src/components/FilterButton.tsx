import { useEffect, useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { completeCount, getIsFiltered, setIsFiltered } from 'src/store/taskSlice'

export const FilterButton = () => {
  const dispatch = useDispatch()
  const isFiltered = useSelector(getIsFiltered)
  const completeCountItems = useSelector(completeCount)

  const [showCompleteTasks, setShowCompleteTasks] = useState(isFiltered)

  useEffect(() => {
    dispatch(setIsFiltered(showCompleteTasks))
  }, [showCompleteTasks])

  const handleClick = () => {
    setShowCompleteTasks(prev => !prev)
  }
  return (
    <div className="filter">
      <button className="filter-button" onClick={handleClick} data-testid="кнопка фильтрации">
        {!showCompleteTasks ? 'Скрыть выполненные задачи' : 'Показать выполненные задачи:'}
      </button>
      {showCompleteTasks && (
        <div className="filter-complete-count" data-testid="количество задач">
          {completeCountItems} шт.
        </div>
      )}
    </div>
  )
}
