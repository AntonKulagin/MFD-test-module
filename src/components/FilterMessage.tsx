import React from 'react'

export const FilterMessage = () => {
  return <div className="filter-message">Нет выполненных задач</div>
}
