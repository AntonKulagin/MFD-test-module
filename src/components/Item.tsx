import { useSelector } from 'react-redux'
import { DeleteButton } from './DeleteButton'
import { uncompleteCount } from 'src/store/taskSlice'

type Props = Task & {
  onDelete: (id: Task['id']) => void
  onToggle: (id: Task['id']) => void
}

export const Item = (props: Props) => {
  const uncompleteCountTasks = useSelector(uncompleteCount)

  const testControlHeader = props.header
  if (!testControlHeader.length || testControlHeader.length > 32) {
    return null
  }

  const handleToggle = (id: string, done: boolean) => {
    if (uncompleteCountTasks >= 10 && done) {
      return
    }
    props.onToggle(id)
  }

  return (
    <li className="item-wrapper">
      <input
        type="checkbox"
        disabled={uncompleteCountTasks >= 10 && props.done}
        id={props.id}
        defaultChecked={props.done}
        onChange={() => handleToggle(props.id, props.done)}
      />
      <label htmlFor={props.id} onClick={() => handleToggle(props.id, props.done)}>
        {props.done ? <s>{props.header}</s> : props.header}
      </label>
      <DeleteButton disabled={!props.done} onClick={() => props.onDelete(props.id)} />
    </li>
  )
}
