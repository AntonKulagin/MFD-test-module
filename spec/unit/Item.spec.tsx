import { render, screen } from '@testing-library/react'
import { Item } from 'src/components/Item'
import * as taskSliceSelectors from 'src/store/taskSlice'
import { JestStoreProvider } from '../utils/JestStoreProvider'

describe('Элемент списка задач', () => {
  it('название не должно быть больше 32 символов', () => {
    const onDelete = jest.fn()
    const onToggle = jest.fn()
    const spied = jest.spyOn(taskSliceSelectors, 'uncompleteCount').mockReturnValue(1)
    const item: Task = {
      id: '1111',
      header: 'Название задачи больше 32 символов ',
      done: false,
    }

    render(<Item {...item} key={item.id} onDelete={onDelete} onToggle={onToggle} />, {
      wrapper: JestStoreProvider,
    })

    const itemLabel = screen.queryByLabelText(/название задачи/i)

    expect(itemLabel).not.toBeInTheDocument()
  })
  it('название не должно быть пустым', () => {
    const onDelete = jest.fn()
    const onToggle = jest.fn()
    const spied = jest.spyOn(taskSliceSelectors, 'uncompleteCount').mockReturnValue(1)
    const item: Task = {
      id: '1111',
      header: '',
      done: false,
    }

    render(<Item {...item} key={item.id} onDelete={onDelete} onToggle={onToggle} />, {
      wrapper: JestStoreProvider,
    })

    const itemLabel = screen.queryByLabelText('')

    expect(itemLabel).not.toBeInTheDocument()
  })
  it('нельзя удалять невыполненные задачи', () => {
    const onDelete = jest.fn()
    const onToggle = jest.fn()
    const spied = jest.spyOn(taskSliceSelectors, 'uncompleteCount').mockReturnValue(1)
    const item: Task = {
      id: '1111',
      header: 'невыполненная задача',
      done: false,
    }

    render(<Item {...item} key={item.id} onDelete={onDelete} onToggle={onToggle} />, {
      wrapper: JestStoreProvider,
    })

    const liElement = screen.getByRole('listitem')

    const itemButton = screen.getByRole('button')

    expect(liElement).toMatchSnapshot()
    expect(itemButton).toBeDisabled()
  })
})
