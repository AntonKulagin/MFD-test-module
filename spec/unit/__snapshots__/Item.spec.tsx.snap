// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Элемент списка задач нельзя удалять невыполненные задачи 1`] = `
<li
  class="item-wrapper"
>
  <input
    id="1111"
    type="checkbox"
  />
  <label
    for="1111"
  >
    невыполненная задача
  </label>
  <button
    class="button button-with-icon"
    data-alt="Удалить"
    disabled=""
  >
    <img
      alt="Удалить"
      src="[object Object]"
    />
  </button>
</li>
`;
